action.yml

```diff
@@ -0,0 +1,266 @@
+    required: false
+    default: 'false'
+
+  google_ar:
+    description: 'Enable Google Artifact Registry'
+    required: false
+    default: 'false'
+
+  google_ar_image_name:
+    description: 'Google Artifact Registry image name'
+    required: false
+    default: ''
+
+  google_ar_tag_prefix:
+    description: 'Google Artifact Registry latest tag'
+    required: false
+    default: ''
+
+  google_workload_identity_provider:
+    description: 'Google Workload Identity Provider'
+    required: false
+    default: ''
+
+  google_service_account:
+    description: 'Google Service Account'
+    required: false
+    default: ''
+
+  # Build configuration
+  architecture:
+    description: 'Architecture to build (e.g., amd64, arm64)'
+    required: false
+    default: 'amd64'
+
+  # Cache configuration
+  cache_enabled:
+    description: 'Enable build cache'
+    required: false
+    default: 'true'
+
+  cache_suffix:
+    description: 'Cache image suffix'
+    required: false
+    default: 'cache'
+
+outputs:
+  image:
+    description: 'Built image tag'
+    value: ${{ steps.output-images.outputs.image }}
+
+  image_digest:
+    description: 'Image digest'
+    value: ${{ steps.output-images.outputs.digest }}
+
+  pushed:
+    description: 'Whether image was pushed to registries'
+    value: ${{ steps.output-images.outputs.pushed }}
+
+runs:
+  using: 'composite'
+  steps:
+    # =============================================================================
+    # BUILD IMAGES
+    # =============================================================================
+
+    - name: Setup build metadata
+      id: setup
+      shell: bash
+      run: |
+        # Setup image names
+        GHCR_IMAGE_NAME="${{ inputs.ghcr_image_name }}"
+        if [[ -z "$GHCR_IMAGE_NAME" ]]; then
+          GHCR_IMAGE_NAME="ghcr.io/${{ github.repository_owner }}/${{ inputs.image_name }}"
+        fi
+        echo "ghcr_image_name=$GHCR_IMAGE_NAME" >> $GITHUB_OUTPUT
+
+        # Use correct SHA for the event type
+        if [[ "${{ github.event_name }}" == "pull_request" ]]; then
+          COMMIT_SHA="${{ github.event.pull_request.head.sha }}"
+        else
+          COMMIT_SHA="${{ github.sha }}"
+        fi
+        echo "commit_sha=$COMMIT_SHA" >> $GITHUB_OUTPUT
+        echo "Event: ${{ github.event_name }}"
+        echo "Branch: ${{ github.ref_name }}"
+        echo "Architecture: ${{ inputs.architecture }}"
+        echo "Commit SHA: $COMMIT_SHA"
+        if [[ "${{ inputs.google_ar }}" == "true" ]]; then
+          if [[ -z "${{ inputs.google_ar_image_name }}" ]]; then
+            echo "Error: Google Artifact Registry image name is required when publshing to Google Artifact Registry is enabled"
+            exit 1
+          fi
+          echo "GAR Image: ${{ inputs.google_ar_image_name }}"
+        fi
+        if [[ "${{ inputs.ghcr }}" == "true" ]]; then
+          echo "GHCR Image: $GHCR_IMAGE_NAME"
+        fi
```

Comment:
Could we use the `docker/metadata-action` here instead (https://github.com/docker/metadata-action)? It gives us a lot of things for free -- like annotations used for provenance and attestations.